app/src/actions/websocket.js

```diff
@@ -0,0 +1,54 @@
+/* global WebSocket */
+import types from './actionTypes'
+import { store } from '../store'
+
+const URI = process.env.REACT_APP_WS || 'ws://localhost:3000'
+let webSocket = null
+
+let pingTimeout = null
+
+const heartbeat = () => {
+  clearTimeout(pingTimeout)
+
+  pingTimeout = setTimeout(() => {
+    webSocket.close()
+  }, 30000 + 1000) // 30 seconds
+}
+
+const check = () => {
+  if (!webSocket || webSocket.readyState === WebSocket.CLOSED) initWebsocket()
+}
+
+const initWebsocket = () => {
+  webSocket = new WebSocket(URI)
+
+  webSocket.onmessage = (event) => {
+    const data = JSON.parse(event.data)
+
+    if (data.message && data.message === 'ping') {
+      webSocket.send(JSON.stringify({ message: 'pong' }))
+      heartbeat()
+      return
+    }
+
+    store.dispatch({ type: types.OPEN_TOAST, payload: { message: data.judgement.msg } })
+  }
+
+  webSocket.onopen = () => {
+    console.log(` Ws: Connected to ${URI}`)
+    heartbeat()
+  }
+  webSocket.onclose = () => {
+    console.log(` Ws: Disconnected from server`)
+    webSocket = null
+    clearTimeout(pingTimeout)
+    check(webSocket, store)
```

Comment:
why is the store passed here? also you already know the state is closed, you might as well reinit right there

Reply:
investigate if websocket offers a way to reconnect without reconstructing the object

Reply:
perhaps it's a better idea to encapsulate this behaviour more clearly rather than abuse globals like this

Reply:
and have a ws object which auto-reconnects which the application will use with peace of mind

Reply:
either this or investigate a websocket library which does auto-reconnect

Reply:
> why is the store passed here? also you already know the state is closed, you might as well reinit right there

My mistake. In a previous version of the code I was passing around the store and I forgot to delete the arguments to reflect the new changes.

> investigate if websocket offers a way to reconnect without reconstructing the object
> perhaps it's a better idea to encapsulate this behaviour more clearly rather than abuse globals like this
> and have a ws object which auto-reconnects which the application will use with peace of mind

WebSocket library does not auto-reconnect. You have to implemented yourself or use a library like sockets.io (which I avoided to use because is too much imo for our needs). Any suggestion how I could make it clearly without abuse module globals ?

Reply:
Btw this auto-reconnect has the problem that it can flood the server, if a high number of clients are connected to the server when it reboots, but I think is ok for the moment.

Reply:
Yes. So in order to:

* hide this implementation detail which is not relevant
* have the reconnection mechanism work in a predictable manner

I suggest you do use a library instead of reimplementing it yourself.

Reply:
If you really want to write it yourself, you should have your own connection object to encapsulate the underlying websocket object.

Reply:
But I don't suggest you do.